### PR TITLE
Support exchange sub partition with external table

### PR DESCRIPTION
**Pull request by fanfuxiaoran**
Signed-off-by: Xiaoran Wang <xiwang@pivotal.io>

### CODE REVIEW

**Comment by adam8157** [comment]
Please add some commit message, such as the syntax, and why is this safe.

About the test, I'm happy to see selecting data from it.

**Comment by fanfuxiaoran** [comment]
Selecting data from the  external table needs to start a gpfdist instance .
Does it need to be added  to  the test ?

**Comment by adam8157** [comment]
`file://` or a external web table works too

**Comment by fanfuxiaoran** [comment]
Updated the pr.
I define a external web table generating data through 'echo' command.

**Comment by yydzero** [comment]
@xiong-gang could you please review this PR?

I am wondering why the original code disable exchange subpartitions with external table.

File: src/test/regress/input/filespace.source [hunk 1]
```diff
@@ -1403,6 +1403,26 @@ set client_min_messages=ERROR;
 drop database dropdb_same_relfilenode;
 reset client_min_messages;
 
+
+-- exchange sub partition with external table
```
Comment (xiong-gang) [inline src/test/regress/input/filespace.source]: why in `filespace.source`?

File: src/backend/commands/tablecmds.c [hunk 2]
```diff
@@ -13278,16 +13278,6 @@ ATPExecPartExchange(AlteredTableInfo *tab, Relation rel, AlterPartitionCmd *pc)
...
-		{
-			if (prule && prule->topRule && prule->topRule->parparentoid)
-			{
-				heap_close(newrel, NoLock);
-				ereport(ERROR,
-						(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
-						 errmsg("cannot exchange sub partition with external table")));
-			}
-		}
 		heap_close(newrel, NoLock);
```
Comment (xiong-gang) [inline src/backend/commands/tablecmds.c]: can also remove L13280 L13291

**Comment by fanfuxiaoran** [comment]
All tests of exchange table are in filespace.source.
Where the test should be ?

**Comment by fanfuxiaoran** [comment]
f partitioned tables that have been altered to use an external table as a leaf child partition, GPORCA doesn't support to query.

https://gpdb.docs.pivotal.io/43140/admin_guide/query/topics/query-piv-opt-limitations.html

File: src/backend/commands/tablecmds.c (hunk 2 above)
Comment (danielgustafsson) [reply src/backend/commands/tablecmds.c]: I would upgrade that statement to "have to" remove..

File: src/backend/commands/tablecmds.c [hunk 3]
```diff
@@ -13278,16 +13278,6 @@ ATPExecPartExchange(AlteredTableInfo *tab, Relation rel, AlterPartitionCmd *pc)
 		newrelid = RangeVarGetRelid(newrelrv, false);
 		Assert(OidIsValid(newrelid));
 		newrel = heap_open(newrelid, NoLock);
-		if (RelationIsExternal(newrel))
```
Comment (danielgustafsson) [inline src/backend/commands/tablecmds.c]: If this is allowed, how will pg_dump/pg_restore support this table hierarchy?

File: src/test/regress/input/filespace.source (hunk 1 above)
Comment (danielgustafsson) [reply src/test/regress/input/filespace.source]: This should be in one of the partitioning tests, not in filespace.

File: src/backend/commands/tablecmds.c (hunk 3 above)
Comment (adam8157) [reply src/backend/commands/tablecmds.c]: Turns out pg_restore doesn't support regular partition as an external table..

https://github.com/greenplum-db/gpdb/issues/3569

**Comment by fanfuxiaoran** [comment]
Move test from filespace to partindex_test .

**Comment by adam8157** [comment]
@BrianPivotal pg_dump/restore/upgrade don't support this, they even don't support external partition now, is it OK for the customer who requested this story?

**Comment by adam8157** [comment]
Confirmed with PM, customer is OK with a known issue of pg_dump/restore/upgrade.

**Pull request merged**